### PR TITLE
added custom format loader guide

### PR DESCRIPTION
**Pull request by hungrymonkey**
added the rest of the minor fixes

Pull Requests should use the master branch by default.

Only make Pull Requests against other branches (e.g. 2.1 or 3.0) if your changes only apply to that specific version of Godot.

### CODE REVIEW

File: development/cpp/custom_resource_format_loaders.rst [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+- `core/io/image_loader.h <https://github.com/godotengine/godot/blob/master/core/io/image_loader.h>`__
+
+
+Creating a ResourceFormatLoader
+-------------------------------
+
+Each file format consist of a data container and a ``ResourceFormatLoader``. 
+
+ResourceFormatLoaders are usually simple classes which returns all the
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: return, not returns

File: development/cpp/custom_resource_format_loaders.rst [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+- `core/io/image_loader.h <https://github.com/godotengine/godot/blob/master/core/io/image_loader.h>`__
+
+
+Creating a ResourceFormatLoader
+-------------------------------
+
+Each file format consist of a data container and a ``ResourceFormatLoader``. 
+
+ResourceFormatLoaders are usually simple classes which returns all the 
+necessary metadata for godot to understanding the new extension. The
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: necessary metadata for **G**odot to understand the new extension (capitalize Godot, understand -> understanding)

File: development/cpp/custom_resource_format_loaders.rst [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+
+Creating a ResourceFormatLoader
+-------------------------------
+
+Each file format consist of a data container and a ``ResourceFormatLoader``. 
+
+ResourceFormatLoaders are usually simple classes which returns all the 
+necessary metadata for godot to understanding the new extension. The 
+class must the return format name and the extension string.
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: the format name

File: development/cpp/custom_resource_format_loaders.rst [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+In addition, ResourceFormatLoaders must convert file paths into 
+resources with the function ``load``. To load a resource, the ``load`` must
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: a resource, the ``load`` **function** must

File: development/cpp/custom_resource_format_loaders.rst [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	};
+	#endif // MY_JSON_LOADER_H
+
+.. code:: cpp
+
+	#include "my_json_loader.h"
+	#include "my_json.h"
+
+	ResourceFormatLoaderMyJson::ResourceFormatLoaderMyJson() {
+	}
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: insert newline after this

File: development/cpp/custom_resource_format_loaders.rst [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+		if (r_error)
+			*r_error = OK;
+		Error err = my->set_file(p_path);
+		return Ref<MyJson>(my);
+	}
+
+	void ResourceFormatLoaderMyJson::get_recognized_extensions(List<String> *p_extensions) const {
+
+		p_extensions->push_back("mjson");
+	}
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: another newline here

File: development/cpp/custom_resource_format_loaders.rst [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+	bool ResourceFormatLoaderMyJson::handles_type(const String &p_type) const {
+		return (p_type == "MyJson");
+	}
+
+
+Creating Custom Data Types
+--------------------------
+
+
+Sometimes, Godot do not have the necessary facilities to express every data format.
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: do = does. Also... I don't really get what this says? Could you expand upon this?

File: development/cpp/custom_resource_format_loaders.rst [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+	class MyJson : public Resource{
+	    GDCLASS(MyJson, Resource);
+
+	protected:
+	    static void _bind_methods() {
+			ClassDB::bind_method(D_METHOD("toString"), &MyJson::toString);
+		}
+	private:
+	    Dictionary dict;
+	public:
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: add newlines after each method's closing brace

File: development/cpp/custom_resource_format_loaders.rst [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+	    MyJson() {};
+	    ~MyJson() {};
+	};
+	#endif
+
+Considerations
+~~~~~~~~~~~~~~
+
+Some libraries uses interfaces such as ``std::istream`` for
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: uses = use

File: development/cpp/custom_resource_format_loaders.rst [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+	};
+	#endif
+
+Considerations
+~~~~~~~~~~~~~~
+
+Some libraries uses interfaces such as ``std::istream`` for 
+deserializing custom data types.
+
+Here is the code for code for translating ``FileAccess`` calls
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: duplicate "code for"

File: development/cpp/custom_resource_format_loaders.rst [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    
+	class GodotFileInStreamBuf : public std::streambuf{
+	public:
+		GodotFileInStreamBuf(FileAccess * fa) {
+			_file = fa;
+		}
+		int underflow(){
+			if (_file->eof_reached()){
+			return EOF;
+		}else{
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: }else{ = } else { (insert spaces)

File: development/cpp/custom_resource_format_loaders.rst [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+	#include <istream>
+	#include <streambuf>
+    
+	class GodotFileInStreamBuf : public std::streambuf{
+	public:
+		GodotFileInStreamBuf(FileAccess * fa) {
+			_file = fa;
+		}
+		int underflow(){
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: insert space before the brace {

File: development/cpp/custom_resource_format_loaders.rst [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+			if (_file->eof_reached()){
+			return EOF;
+		}else{
+			size_t pos = _file->get_position();
+			uint8_t ret = _file->get_8();
+			_file->seek(pos); //required since get_8() advances the read head
+			return ret;
+			}
+		}
+		int uflow(){
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: insert space before the brace {

File: development/cpp/custom_resource_format_loaders.rst [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+~~~~~~~~~~~
+- `istream <http://www.cplusplus.com/reference/istream/istream/>`__
+- `streambuf <http://www.cplusplus.com/reference/streambuf/streambuf/?kw=streambuf>`__
+- `core/io/fileaccess.h <https://github.com/godotengine/godot/blob/master/core/os/file_access.h>`__
+
+Registering the New File Format
+-------------------------------
+
+Godot registers ResourcesFormatLoader with a ResourceLoader 
+handler. The handler will select the proper loader automatically
```
Comment (mhilbrunner) [inline development/cpp/custom_resource_format_loaders.rst]: missing "." after "automatically"

**Comment by hungrymonkey** [comment]
So many mistakes. I was tried when I wrote this

File: development/cpp/custom_resource_format_loaders.rst (hunk 7 above)
Comment (hungrymonkey) [reply development/cpp/custom_resource_format_loaders.rst]: Godot does not have the proper data types for things like ML models. In order to expose ML models to GDscript, users must create custom classes,

Comment (hungrymonkey) [reply development/cpp/custom_resource_format_loaders.rst]: I am going to rewrite that part then. I was a bit too tired

**Comment by mhilbrunner** [comment]
No worries, happens to the best of us :)

**Comment by hungrymonkey** [comment]
one down. one more to go. i have to write about those godot server stuff

**Pull request merged**